addons/mass_mailing/controllers/main.py

```diff
@@ -167,23 +172,16 @@ def view(self, mailing_id, email=None, res_id=None, token=""):
             if not self._valid_unsubscribe_token(mailing_id, res_id, email, str(token)) and not request.env.user.has_group('mass_mailing.group_mass_mailing_user'):
                 raise exceptions.AccessDenied()
 
-            res = mailing.convert_links()
-            base_url = mailing.get_base_url().rstrip('/')
-            urls_to_replace = [
-                (base_url + '/unsubscribe_from_list', mailing._get_unsubscribe_url(email, res_id)),
-                (base_url + '/view', mailing._get_view_url(email, res_id))
-            ]
-            for url_to_replace, new_url in urls_to_replace:
-                if url_to_replace in res[mailing_id]:
-                    res[mailing_id] = res[mailing_id].replace(url_to_replace, new_url if new_url else '#')
-
-            res[mailing_id] = res[mailing_id].replace(
-                'class="o_snippet_view_in_browser"',
-                'class="o_snippet_view_in_browser" style="display: none;"'
-            )
+            html = str(request.env['mail.render.mixin'].sudo()._render_template_qweb(
+                mailing.body_html, mailing.mailing_model_real, [res_id], add_context=None, options={})[res_id])
+            # Update generic URLs (without parameters) to final ones
+            html = html.replace('/unsubscribe_from_list', mailing._get_unsubscribe_url(email, res_id).translate(
```

Comment:
no need to create a variable for this imo
```python
unsub_url = mailing._get_unsubscribe_url(email, res_id)
unsub_url = unsub_url.translate({'"': '', '<': '', '>': '', "'": ''})
html = html.replace('/unsubscribe_from_list', unsub_url)**
```

btw, can't we escape them instead of removing them ?

Reply:
As suggested, I have used urllib.parse.quote but added some safe characters in order to not escape/encode important characters of the URL (ex.: % already used to encode the parameters).

Note: it seems you cannot use a dictionary as parameter for translate method, you have to convert it with "str.maketrans" otherwise it does nothing ...